### DISCUSSION TITLE
How did you figure out how to store dates as strings in sqlite?

### DISCUSSION THREAD

**Discussion by dschissler** [pocketbase/pocketbase#1067] [Q&A] [answered]
When I was starting out a bit ago I wasn't able to figure out how to store dates as standard strings in sqlite because of something to do with the serializers / converters or something.  It was said that there was some kind of ambiguity all of the way into the sqlite interface / drivers.  I was just starting out and so that wasn't fun to be hit like that.  I was shocked that Gorm and others ORMS were just throwing their hands up in the air about it, and then proposing that we could possibly add some of our own methods.  So in the end I went with dates stored as integer Unix time.

How did you figure this thing out?  Are you doing special conversions or did some new fixes come through?  At any rate, I faced some real pitfalls before that make me really appreciate Pocketbase.

**Comment by ganigeorgiev** [comment, marked as answer]
SQLite types are fairly generic and even allows storing mixed values types in the same column (I think strict table definitions were added recently an year ago). With that said there are no dedicated datetime/timestamp/etc. types and you can store the datetime strings as text, real or integer (https://www.sqlite.org/datatype3.html#date_and_time_datatype).

For PocketBase I've decided to represent dates as `TEXT` for several reasons:
- it is easier to read and in some simple cases can be displayed directly without additional formatting
- it makes the date filters more human friendly and even allow "non-full/partial" date lookups (eg. `created > "2022-01-01"`; in the future there are plans to make the filter context aware and we may support both integer and text when searching)
- it will be easier to migrate the database of my other open source project [Presentator #183](https://github.com/presentator/presentator/issues/183) (that's why we also use bcrypt instead of newer algos like argon for the passwords)

Creating custom db types is not really that difficult in Go. You'll need to implement mainly 2 interfaces- [sql.Scanner](https://pkg.go.dev/database/sql#Scanner) and [driver.Valuer](https://pkg.go.dev/database/sql/driver#Valuer). For convenience, PocketBase comes with its own `types.DateTime` struct (internally it uses `time.Time`) - https://github.com/pocketbase/pocketbase/blob/master/tools/types/datetime.go.

**Comment by dschissler** [comment]
> Creating custom db types is not really that difficult in Go. You'll need to implement mainly 2 interfaces- [sql.Scanner](https://pkg.go.dev/database/sql#Scanner) and [driver.Valuer](https://pkg.go.dev/database/sql/driver#Valuer).

It may not be that difficult to read the code, but when you're just starting out with a completely new language and stack then it's quite brutal.  For me it's just not a level that I want to be playing with until deep into a project when it becomes relevant for some reason like efficiency or some trick moves.
